### PR TITLE
New version of Maestro API with support for new build information

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: https://github.com/dotnet/arcade/issues/1799

*Notes:*

- Tested this locally with a copy of the production Builds table.

- Maestro.Data/Build.cs : added new properties for azdo/github information. Decided not to use owned types because they only accept nullable types if [stored in a separate table](https://docs.microsoft.com/en-us/ef/core/modeling/owned-entities#limitations) (weird) and I didn't think it was worth to store only two fields: githubbranch and githubrepository.

- Migrations/20190117231810_RefactorBuildEntity.cs : Added SQL to the Up & Down methods so that they move existing data from the repository/branch columns to {Github, Azdo}x{Repository, Branch} new columns.

- v2019_01_16/Controllers/BuildsController.cs : Added new controller for handling BuildData with new properties. I had to reimplement these methods because using the base class methods would not return the new properties added to BuildData. The *Create* method had to be explicitly removed from the Api because it couldn't be just overridden since it uses BuildData class from previous version as parameter.

- I patched the old/new API to try to make the right thing when accessing data pushed by new/old API version.

### CODE REVIEW

File: src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs [hunk 1]
```diff
@@ -83,6 +83,7 @@ public class BuildAssetRegistryContext : IdentityDbContext<ApplicationUser, Iden
         protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
         {
             optionsBuilder.AddDotNetExtensions();
+            optionsBuilder.EnableSensitiveDataLogging();
```
Comment (JohnTortugo) [inline src/Maestro/Maestro.Data/BuildAssetRegistryContext.cs]: I'll remove this.

File: src/Maestro/Maestro.Data/Models/Build.cs [hunk 2]
```diff
@@ -34,14 +35,26 @@ static Build()
...
+        public string Commit { get; set; }
 
-        public string Repository { get; set; }
+        public int AzureDevOpsBuildId { get; set; }
 
-        public string Branch { get; set; }
+        public int AzureDevOpsBuildDefinitionId { get; set; }
 
-        public string Commit { get; set; }
+        public string AzureDevOpsAccount { get; set; }
```
Comment (alexperovich) [inline src/Maestro/Maestro.Data/Models/Build.cs]: Can we use Owned Types for the AzDev part of this class? I'd like to organize it that way.

File: src/Maestro/Maestro.Web/Api/v2018_07_16/Models/BuildData.cs [hunk 3]
```diff
@@ -27,13 +29,18 @@ public class BuildData
 
         public Data.Models.Build ToDb()
         {
+            var isGitHubInfo = Repository.IndexOf("github", StringComparison.OrdinalIgnoreCase) >= 0;
+
             return new Data.Models.Build
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Api/v2018_07_16/Models/BuildData.cs]: This needs to save off the BuildNumber also. Even if the build is from a github repository it will still have an AzureDevOpsBuildNumber.

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Models/Build.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using JetBrains.Annotations;
+
+namespace Maestro.Web.Api.v2019_01_16.Models
+{
+    public class Build : Maestro.Web.Api.v2018_07_16.Models.Build
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Api/v2019_01_16/Models/Build.cs]: I think this model shouldn't inherit the previous. You need to remove some of the fields.

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Models/Build.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+            AzureDevOpsBuildNumber = other.AzureDevOpsBuildNumber;
+            AzureDevOpsRepository = other.AzureDevOpsRepository;
+            AzureDevOpsBranch = other.AzureDevOpsBranch;
+        }
+
+        public int AzureDevOpsBuildId { get; set; }
+
+        public int AzureDevOpsBuildDefinitionId { get; set; }
+
+        public string AzureDevOpsAccount { get; set; }
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Api/v2019_01_16/Models/Build.cs]: Can you move these into AzureDevOpsInfo and GitHubInfo properties so the properties are grouped correctly.

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Models/BuildData.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using Maestro.Web.Api.v2018_07_16.Models;
+using Microsoft.AspNetCore.ApiVersioning;
+
+namespace Maestro.Web.Api.v2019_01_16.Models
+{
+    public class BuildData : Maestro.Web.Api.v2018_07_16.Models.BuildData
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Api/v2019_01_16/Models/BuildData.cs]: This one should probably also not inherit the previous because you need to remove properties.

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Models/BuildData.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        public string AzureDevOpsProject { get; set; }
+
+        [Required]
+        public string AzureDevOpsBuildNumber { get; set; }
+
+        [Required]
+        public string AzureDevOpsRepository { get; set; }
+
+        [Required]
+        public string AzureDevOpsBranch { get; set; }
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Api/v2019_01_16/Models/BuildData.cs]: And please group these also.

File: src/Maestro/Maestro.Web/Startup.Api.cs [hunk 8]
```diff
@@ -41,6 +41,8 @@ private void ConfigureApiServices(IServiceCollection services)
             services.AddSwaggerGen(
                 options =>
                 {
+                    options.CustomSchemaIds(x => x.FullName);
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Startup.Api.cs]: Why did you add this?

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Models/Build.cs (hunk 4 above)
Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Api/v2019_01_16/Models/Build.cs]: Great. Will change it.

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Models/BuildData.cs (hunk 6 above)
Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Api/v2019_01_16/Models/BuildData.cs]: I'll change it.

File: src/Maestro/Maestro.Data/Models/Build.cs (hunk 2 above)
Comment (JohnTortugo) [reply src/Maestro/Maestro.Data/Models/Build.cs]: Are you fine with using Owned Types just for AzDev? I thought it would be strange to use owned types only for AzDev and leave GitHub info on the Build property. The overall code is also much simpler if not using owned types.

File: src/Maestro/Maestro.Web/Startup.Api.cs (hunk 8 above)
Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Startup.Api.cs]: Without this I get an error saying that I have a duplicated type Build, because by default swagger doesn't consider namespaces on the type name. Error is this: `Identical schemaIds detected for types Maestro.Web.Api.v2019_01_16.Models.Build and Maestro.Web.Api.v2018_07_16.Models.Build`. See comment [here](https://stackoverflow.com/a/47306578/973795) also.

Comment (alexperovich) [reply src/Maestro/Maestro.Web/Startup.Api.cs]: This shouldn't be necessary, the old Build type should not be included in the new api. If it is thats a bug.

Comment (JohnTortugo) [reply src/Maestro/Maestro.Web/Startup.Api.cs]: Fix included in next push.

**Comment by JohnTortugo** [comment]
Can someone PTAL?

Would making the BuildInfo objects owned types be a blocking issue?

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/BuildsController.cs [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+{
+    public class BuildsController_ApiRemoved : Maestro.Web.Api.v2018_07_16.Controllers.BuildsController
+    {
+        public BuildsController_ApiRemoved(BuildAssetRegistryContext context)
+            : base(context)
+        {
+        }
+
+        [ApiRemoved]
+        public override sealed IActionResult GetAllBuilds(
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/BuildsController.cs]: Technically for most of these you don't need to "ApiRemoved" them. If the new action method has the same parameters and return type as the old method you can just override it and write the new implementation.

You also don't need a separate class to contain the [ApiRemoved] things.

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/SubscriptionsController.cs [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+        public SubscriptionsController_ApiRemoved(
+            BuildAssetRegistryContext context,
+            BackgroundQueue queue,
+            IDependencyUpdater dependencyUpdater,
+            Func<ActorId, ISubscriptionActor> subscriptionActorFactory)
+            : base(context, queue, dependencyUpdater, subscriptionActorFactory)
+        {
+        }
+
+        [ApiRemoved]
```
Comment (alexperovich) [inline src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/SubscriptionsController.cs]: Same thing here, if the signatures match just override without the [ApiRemoved] thing.

**Comment by JohnTortugo** [comment]
Thanks @alexperovich . I'll address your comments. I'll wait until we have build preview out to merge this PR.

File: src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/BuildsController.cs (hunk 9 above)
Comment (alexperovich) [reply src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/BuildsController.cs]: Thats odd, you should just be able to override and change the [SwaggerResponse] attribute, it shouldn't need the [HttpGet] attribute again.

**Comment by JohnTortugo** [comment]
Talked with @alexperovich on Teams. He's OK with merging this as is.
